feat: New Grouping docs (#2173)
getsentry/sentry-docs: src/platforms/common/data-management/event-grouping/grouping-enhancements.mdx

> Here is a practical example to see how this looks:

Reviewer: ```suggestion
The following is a practical example of how this looks:
```

```discover {tabTitle: Enhancement Rules}
# mark all functions in the std namespace to be outside the app
family:native stack.function:std::*       -app

# mark all code in node modules not to be in app
stack.abs_path:**/node_modules/**         -app

# remove all generated javascript code from all grouping
stack.abs_path:**/generated/**.js         -group
```

## Matchers

> The following **matchers** exist. Multiple matchers can be defined in a line:

Reviewer: ```suggestion
These **matchers** are available; multiple matchers can be defined in a line:
```

<DefinitionList>

### `family`

> matches on the general platform family. Right now there are `javascript`,

Reviewer: ```suggestion
Matches on the general platform family, which currently includes `javascript`,
```

> `native` and `other`. To make a rule apply to multiple platforms, you can comma

Reviewer: ```suggestion
`native` and `other`. Comma separate rules to apply them to multiple platforms.
```

> separate them.

Reviewer: delete this line (included above)

```discover {tabTitle:Enhancement Rules}
family:javascript,native stack.abs_path:**/generated/**  -group
```

### `stack.abs_path`

alias: `path`

> This matches case insensitive with Unix glob behavior on a path in a stacktrace.

Reviewer: ```suggestion
This matches case insensitive with Unix glob behavior on a path in a stack trace.
```

The path separators are normalized to `/`. As a special rule, if the filename is
relative, it still matches on `**/`.

```discover {tabTitle:Enhancement Rules}
# match on all files under `project` with a `.c` extension
stack.abs_path:**/project/**.c` +app

# matches on vendor/foo without sub folders
stack.abs_path:**/vendor/foo/*.c` -app

# matches on `foo.c` as well as `foo/bar.c`.
stack.abs_path:**/*.gen.c` -group
```

### `stack.module`

alias: `module`

> Module is similar to paths but matches on the module. This is not used for Native,

Reviewer: ```suggestion
Module is similar to `path` but matches on the `module`. This is not used for Native,
```

but it is used for JavaScript, Python, and similar platforms. Matches are case

> sensitive, and normal globbing is available. Note that modules are not packages

Reviewer: ```suggestion
sensitive, and normal globbing is available. Note that modules are not packages,
```

which can be confusing for native environments.

### `stack.function`

alias: `function`

> Matches on a function in a stacktrace, and is case sensitive with normal globbing.

Reviewer: ```suggestion
Matches on a function in a stack trace, and is case sensitive with normal globbing.
```

```discover {tabTitle:Enhancement Rules}
stack.function:myproject_* +app
stack.function:malloc      -group
```

### `stack.package`

alias: `package`

> Matches on a package in a stacktrace. The package is the container that contains a function or module.

Reviewer: ```suggestion
Matches on a package in a stack trace. The package is the container that contains a function or module.
```

This is a `.jar`, a `.dylib` or similar. The same matching rules as for `path` apply (For example, this is typically an absolute path).

```discover {tabTitle:Enhancement Rules}
stack.package:**/libcurl.dylib -group
```

### `app`

> matches on the current state of the in-app flag of a stacktrace frame. `yes`

Reviewer: ```suggestion
Matches on the current state of the in-app flag of a stack trace frame. `yes`
```

means the frame is in app, `no` means it's not.

</DefinitionList>

## Actions

There are two types of **actions**: flag setting and setting variables.

> - **flag**: flags are what is to be done if all matchers match. A flag needs to be prefixed with `+` to set it or `-` to unset it. If this expression is prefixed with a `^`, it applies to frames above the frame (towards the crash). If prefixed with `v` it applies to frames below the frame (away from the crash). For instance, `-group ^-group` removes the matching frame and all frames above it from the grouping.

Reviewer: ```suggestion
- **flag**: flags identify the action to be taken if all matchers match. A flag needs to be prefixed with `+` to set it or `-` to unset it. If this expression is prefixed with a `^`, it applies to frames above the frame (toward the crash). If prefixed with `v` it applies to frames below the frame (away from the crash). For instance, `-group ^-group` removes the matching frame and all frames above it from the grouping.
```

If a line is prefixed with a hash (`#`), it's a comment and ignored.

```discover {tabTitle:Enhancement Rules}
stack.abs_path:**/node_modules/** -group
stack.abs_path:**/app/utils/requestError.jsx -group
stack.abs_path:**src/getsentry/src/getsentry/** +app

family:native max-frames=3

stack.function:fetchSavedSearches v-group
stack.abs_path:**/app/views/**.jsx stack.function:fetchData ^-group

family:native stack.function:SpawnThread v-app -app
family:native stack.function:_NSRaiseError ^-group -app
family:native stack.function:std::* -app
family:native stack.function:core::* -app
```

## Recommendations

> There are some general recommendations we have to greatly improve the out of the box grouping experience.

Reviewer: ```suggestion
These recommendations will greatly improve your out of the box grouping experience.
```

### Mark in-app Frames

To proactively improve your experience, help Sentry determine which frames in your stack trace are "in-app" (part of your own application) and which ones are not. The default rules are defined by the SDK, but in many cases, this can be improved on the server as well. In particular, for languages where server-side processing is necessary (for example, Native C, C++, or JavaScript), it's better to override this on the server.

For instance, the following marks as in-app all frames that are below a specific C++ namespace:

```discover {tabTitle:Enhancement Rules}
stack.function:myapplication::* +app
```

You can also achieve the same result by marking other frames "not in-app." However, if that's the case, you should ensure that first all frames are set to "in-app" to override the defaults:

```discover {tabTitle:Enhancement Rules}
app:no                   +app
stack.function:std::*    -app
stack.function:boost::*  -app
```

> Forcing all frames to be in-app first is necessary because there might already have

Reviewer: ```suggestion
You need to force all frames to be in-app first because there might already have
```

Likewise, you can also chop off the base of a stack trace. This is particularly useful if you have different main loops that drive an application:

```discover {tabTitle:Enhancement Rules}
stack.function:myapp::LinuxMainLoop         v-group -group
stack.function:myapp::MacMainLoop           v-group -group
stack.function:myapp::WinMainLoop           v-group -group
```

### Stack Trace Frame Limits

This isn't useful for _all_ projects, but it can work well for large applications with many crashes. The default strategy is to consider most of the stack trace relevant for grouping. This means that every different stack trace that leads to a crashing function will cause a different group to be created. If you do not want that, you can alternatively force the groups to be much larger by limiting how many frames should be considered.

For instance, if any of the frames in the stack trace refer to a common external library, you could tell the system to only consider the top N frames, :

```discover {tabTitle:Enhancement Rules}

> # always only consider the top 1 frame for all native events

Reviewer: This comment is a bit confusing. I think it means "consistently (or permanently) consider only the top 1 frame for all native events"
